iterative/utils/analytics.go

```diff
@@ -0,0 +1,209 @@
+		t1, _ := time.Parse(layout, matches[len(matches)-1])
+		t0, _ := time.Parse(layout, matches[0])
+		taskDuration = t1.Sub(t0).Seconds()
+	}
+
+	return taskDuration
+}
+
+func IsCI() bool {
+	if _, ok := os.LookupEnv("CI"); ok {
+		return true
+	}
+
+	return false
+}
+
+func TerraformVersion() string {
+	var out bytes.Buffer
+	cmd := exec.Command("terraform", "--version")
+	cmd.Stdout = &out
+
+	err := cmd.Run()
+
+	if err == nil {
+		regex := regexp.MustCompile(`v\d+\.\d+\.\d+`)
+		version := regex.FindString(out.String())
+		return version
+	} else {
+		logrus.Error("Analytics: Failed extracting terraform version")
+	}
+
+	return ""
+}
+
+func GroupId() string {
+	if !IsCI() {
+		return ""
+	}
+
+	id := idgen(fmt.Sprintf("%s%s%s%s%s",
+		os.Getenv("GITHUB_SERVER_URL"),
+		os.Getenv("GITHUB_REPOSITORY_OWNER"),
+		os.Getenv("CI_SERVER_URL"),
+		os.Getenv("CI_PROJECT_ROOT_NAMESPACE"),
+		os.Getenv("BITBUCKET_WORKSPACE"),
+	))
+
+	return id
+}
+
+func UserId() string {
+	id := uuid.New().String()
+	old := appdirs.UserConfigDir("dvc/user_id", "iterative", "", false)
+	new := appdirs.UserConfigDir("telemetry", "iterative", "", false)
+
+	_, errorOld := os.Stat(old)
+	if !os.IsNotExist(errorOld) {
+		os.Rename(old, new)
+	}
+
+	_, errorNew := os.Stat(new)
+	if !os.IsNotExist(errorNew) {
+		dat, _ := ioutil.ReadFile(new)
+		id = string(dat[:])
+	} else {
+		ioutil.WriteFile(new, []byte(id), 0644)
+	}
+
+	if IsCI() {
+		id = idgen(fmt.Sprintf("%s%s%s",
+			os.Getenv("GITHUB_ACTOR"),
+			os.Getenv("GITLAB_USER_ID"),
+			os.Getenv("BITBUCKET_STEP_TRIGGERER_UUID"),
+		))
+	}
+
+	return id
+}
+
+func ResourceData(d *schema.ResourceData) map[string]interface{} {
+	if d == nil {
+		return map[string]interface{}{}
+	}
+
+	tpiLogs := d.Get("logs").([]interface{})
+	logs := ""
+	for _, log := range tpiLogs {
+		logs += log.(string)
+	}
+	spot := d.Get("spot").(float64)
+	return map[string]interface{}{
+		"cloud":           d.Get("cloud").(string),
+		"cloud_region":    d.Get("region").(string),
+		"cloud_machine":   d.Get("machine").(string),
+		"cloud_disk_size": d.Get("disk_size").(int),
+		"cloud_spot":      spot,
+		"cloud_spot_auto": spot == 0.0,
+		"task_status":     d.Get("status").(map[string]interface{}),
+		"task_duration":   TaskDuration(logs),
+		"task_resumed":    len(logs) > 1,
+	}
+}
+
+func JitsuEventPayload(action string, e error, d *schema.ResourceData) map[string]interface{} {
+	systemInfo := SystemInfo()
+
+	extra := ResourceData(d)
+	extra["ci"] = IsCI()
+	extra["terraform_version"] = TerraformVersion()
+
+	err := ""
+	if e != nil {
+		err = e.Error()
+	}
+
+	payload := map[string]interface{}{
+		"user_id":      UserId(),
+		"group_id":     GroupId(),
+		"action":       action,
+		"interface":    "cli",
+		"tool_name":    "tpi",
+		"tool_source":  "terraform",
+		"tool_version": Version,
+		"os_name":      systemInfo["plattform"],
+		"os_version":   systemInfo["platform_version"],
+		"backend":      extra["cloud"],
+		"error":        err,
```

Comment:
Potentially sensitive?

Reply:
error? 
maybe... I think that we hide all the sensitive data within terraform but not sure at what level it is

Reply:
error is defined in the specs

Reply:
For example in AWS this may contain autorization errors in plain text (i.e. after STS decoding them)

Reply:
Should we avoid them then?

Reply:
> _error is defined in the specs_

Yes, and explicitly _“must be stripped of sensitive info e.g. `str(type(exception))`”_

Reply:
> _Should we avoid them then?_

No, just sanitize them. Although Go error management model makes this a bit difficult, to put it mildly.

Reply:
so replace it with `string(type(error)))`?

Reply:
> No, just sanitize them. Although Go error management model makes this a bit difficult, to put it mildly.

Can you please elaborate what do you have in mind?

Reply:
`reflect.TypeOf(err).Name()` would be something like `RequestError` for SDK errors (depends on the provider)

Reply:
If we want to provide more insightful information, we can either send the whole logs (handle with care) or try to extract anonymized information from error messages; for example, match some strings.